<launch>
  <!-- Color mode -->
  <arg name="color_raw"         default="0" />
  <arg name="color_rectified"   default="1" />
  <!-- Depth mode -->
  <arg name="depth_raw"         default="0" />
  <arg name="depth_gray"        default="1" />
  <arg name="depth_colorful"    default="2" />
  <!-- Stream mode -->
  <arg name="stream_640x480"    default="0" />
  <arg name="stream_1280x480"   default="1" />
  <arg name="stream_1280x720"   default="2" />
  <arg name="stream_2560x720"   default="3" />
  <!-- Device mode -->
  <arg name="device_color"      default="0" />
  <arg name="device_depth"      default="1" />
  <arg name="device_all"        default="2" />
  <!-- Stream format -->
  <arg name="stream_mjpg"       default="0" />
  <arg name="stream_yuyv"       default="1" />
  <!-- Process mode -->
  <arg name="proc_none"           default="0" />
  <arg name="proc_imu_assembly"   default="1" />
  <arg name="proc_imu_temp_drift" default="2" />
  <arg name="proc_imu_all"        default="3" />

  <!-- Camera Params -->

  <!-- Device index -->
  <arg name="dev_index" default="0" />
  <!-- Framerate -->
  <arg name="framerate" default="30" />

  <arg name="color_mode" default="$(arg color_rectified)" />
  <arg name="depth_mode" default="$(arg depth_raw)" />
  <arg name="stream_mode" default="$(arg stream_1280x480)" />
  <arg name="device_mode" default="$(arg device_all)" />
  <arg name="color_stream_format" default="$(arg stream_yuyv)" />
  <arg name="depth_stream_format" default="$(arg stream_yuyv)" />

  <!-- Auto-exposure -->
  <arg name="state_ae" default="false" />
  <!-- Auto-white balance -->
  <arg name="state_awb" default="false" />
  <!-- IR intensity -->
  <arg name="ir_intensity" default="4" />

  <!-- Process mode, e.g. imu assembly, temp_drift -->
  <arg name="proc_mode" default="$(arg proc_none)" />

  <!-- Generating points frequency, make slower than framerate -->
  <arg name="points_frequency" default="2" />
  <!-- Points display z distance scale factor -->
  <arg name="points_factor" default="1000.0" />

  <!-- Setup your local gravity here -->
  <arg name="gravity" default="9.8" />

  <!-- Node params -->

  <arg name="mynteye"       default="mynteye" />
  <arg name="base_frame"    default="$(arg mynteye)_link_frame" />

  <arg name="left_mono_frame"   default="$(arg mynteye)_left_mono_frame" />
  <arg name="left_color_frame"  default="$(arg mynteye)_left_color_frame" />
  <arg name="right_mono_frame"  default="$(arg mynteye)_right_mono_frame" />
  <arg name="right_color_frame" default="$(arg mynteye)_right_color_frame" />
  <arg name="depth_frame"   default="$(arg mynteye)_depth_frame" />
  <arg name="points_frame"  default="$(arg mynteye)_points_frame" />
  <arg name="imu_frame"     default="$(arg mynteye)_imu_frame" />
  <arg name="temp_frame"    default="$(arg mynteye)_temp_frame" />

  <!-- left topics -->
  <arg name="left_mono_topic"  default="$(arg mynteye)/left/image_mono" />
  <arg name="left_color_topic" default="$(arg mynteye)/left/image_color" />
  <!-- right topics -->
  <arg name="right_mono_topic"  default="$(arg mynteye)/right/image_mono" />
  <arg name="right_color_topic" default="$(arg mynteye)/right/image_color" />
  <!-- depth topic -->
  <arg name="depth_topic"   default="$(arg mynteye)/depth/image_raw" />
  <!-- points topic -->
  <arg name="points_topic"  default="$(arg mynteye)/points/data_raw" />
  <!-- imu topic -->
  <arg name="imu_topic"     default="$(arg mynteye)/imu/data_raw" />
  <!-- temp topic -->
  <arg name="temp_topic"    default="$(arg mynteye)/temp/data_raw" />

  <node name="mynteye_wrapper_d_node" pkg="mynteye_wrapper_d" type="mynteye_wrapper_d_node" output="screen">

    <!-- Camera params -->
    <param name="dev_index"     value="$(arg dev_index)" />
    <param name="framerate"     value="$(arg framerate)" />
    <param name="color_mode"    value="$(arg color_mode)" />
    <param name="depth_mode"    value="$(arg depth_mode)" />
    <param name="stream_mode"   value="$(arg stream_mode)" />
    <param name="device_mode"   value="$(arg device_mode)" />
    <param name="color_stream_format" value="$(arg color_stream_format)" />
    <param name="depth_stream_format" value="$(arg depth_stream_format)" />
    <param name="state_ae"      value="$(arg state_ae)" />
    <param name="state_awb"     value="$(arg state_awb)" />
    <param name="ir_intensity"  value="$(arg ir_intensity)" />
    <param name="proc_mode"     value="$(arg proc_mode)" />

    <param name="points_factor"    value="$(arg points_factor)" />
    <param name="points_frequency" value="$(arg points_frequency)" />

    <param name="gravity" value="$(arg gravity)" />

    <!-- Frame ids -->
    <param name="base_frame"   value="$(arg base_frame)" />
    <param name="left_mono_frame"   value="$(arg left_mono_frame)" />
    <param name="left_color_frame"  value="$(arg left_color_frame)" />
    <param name="right_mono_frame"  value="$(arg right_mono_frame)" />
    <param name="right_color_frame" value="$(arg right_color_frame)" />
    <param name="depth_frame"  value="$(arg depth_frame)" />
    <param name="points_frame" value="$(arg points_frame)" />
    <param name="imu_frame"    value="$(arg imu_frame)" />
    <param name="temp_frame"   value="$(arg temp_frame)" />

    <!-- Topic names -->

    <param name="left_mono_topic"   value="$(arg left_mono_topic)" />
    <param name="left_color_topic"  value="$(arg left_color_topic)" />
    <param name="right_mono_topic"  value="$(arg right_mono_topic)" />
    <param name="right_color_topic" value="$(arg right_color_topic)" />
    <param name="depth_topic"       value="$(arg depth_topic)" />
    <param name="points_topic"      value="$(arg points_topic)" />
    <param name="imu_topic"         value="$(arg imu_topic)" />
    <param name="temp_topic"        value="$(arg temp_topic)" />
  </node>

  <arg name="pi"   value="3.1415926535897932" />
  <arg name="pi/2" value="1.5707963267948966" />
  <arg name="optical_rotate" value="0 0 0 -$(arg pi/2) 0 -$(arg pi/2)" />

  <node pkg="tf" type="static_transform_publisher" name="b2l_m_broadcaster"
      args="$(arg optical_rotate) $(arg base_frame) $(arg left_mono_frame) 100" />
  <node pkg="tf" type="static_transform_publisher" name="b2l_c_broadcaster"
      args="$(arg optical_rotate) $(arg base_frame) $(arg left_color_frame) 100" />

  <node pkg="tf" type="static_transform_publisher" name="b2r_m_broadcaster"
      args="$(arg optical_rotate) $(arg base_frame) $(arg right_mono_frame) 100" />
  <node pkg="tf" type="static_transform_publisher" name="b2r_c_broadcaster"
      args="$(arg optical_rotate) $(arg base_frame) $(arg right_color_frame) 100" />

  <node pkg="tf" type="static_transform_publisher" name="b2d_roadcaster"
      args="$(arg optical_rotate) $(arg base_frame) $(arg depth_frame) 100" />
  <node pkg="tf" type="static_transform_publisher" name="b2p_broadcaster"
      args="$(arg optical_rotate) $(arg base_frame) $(arg points_frame) 100" />

  <node pkg="tf" type="static_transform_publisher" name="b2imu_broadcaster"
      args="0 0 0 $(arg pi/2) 0 $(arg pi) $(arg base_frame) $(arg imu_frame) 100" />
  <node pkg="tf" type="static_transform_publisher" name="b2temp_broadcaster"
      args="0 0 0 $(arg pi/2) 0 $(arg pi) $(arg base_frame) $(arg temp_frame) 100" />

  <!-- disable compressed depth plugin for image topics -->
  <group ns="$(arg left_mono_topic)">
    <rosparam param="disable_pub_plugins">
      - 'image_transport/compressedDepth'
    </rosparam>
  </group>
  <group ns="$(arg left_color_topic)">
    <rosparam param="disable_pub_plugins">
      - 'image_transport/compressedDepth'
    </rosparam>
  </group>
  <group ns="$(arg right_mono_topic)">
    <rosparam param="disable_pub_plugins">
      - 'image_transport/compressedDepth'
    </rosparam>
  </group>
  <group ns="$(arg right_color_topic)">
    <rosparam param="disable_pub_plugins">
      - 'image_transport/compressedDepth'
    </rosparam>
  </group>
</launch>
